javalib/src/main/scala/java/lang/AbstractStringBuilder.scala

```diff
@@ -489,16 +489,29 @@ abstract class AbstractStringBuilder private (unit: Unit) {
   }
 
   override def toString(): String = {
-    if (count == 0) {
-      return ""
-    }
+    if (count == 0) return ""
+
+    // --- Optimize String sharing for more performance --- //
+    // Disable "sharing" if the used content is considered small compared to the backing Array
+    // See: https://github.com/scala-native/scala-native/issues/2905
+
+    // Estimate the number of wasted characters (placeholders in the backing Array)
     val wasted = value.length - count
-    if (wasted >= 256 ||
-        (wasted >= INITIAL_CAPACITY && wasted >= (count >> 1))) {
+
+    // Case 1: "sharing" disabled
+    // IF the number of wasted characters is higher than the INITIAL_CAPACITY (16)
+    // AND IF the backing Array (value) contains more placeholder characters (wasted)
+    // than what is added by a single enlargeBuffer() operation
+    // (it may happen if setLength() has been previously called to shrink the number of used characters) 
+    // => copy backing Array in a fresh String
+    if (wasted >= INITIAL_CAPACITY && wasted >= (count >> 1) + 2) { // see enlargeBuffer() for details
```

Comment:
I have been studying this line for hours, my limitation, and have been trying to figure
out what data would be needed to make a data driven decision.  Sorry, but I need
more time. I do not want my limited understanding to stand in the way of progress.

I suspect that this line is problematic.

The 256 'magic number' in the original code was never explained.  By reading  of
the original code, the clause to the left of the `||` prevented "huge" amounts of
'wasted' space from getting passed to a String constructor.  That is, it establishes
256 as the maximum amount per String of potentially wasted space of potentially long duration.    One can talk about or, better yet, measure the effect of values smaller than 256 (I suspect 128 or 64 would be better).

By my, possibly flawed reading, this line of code will allow "huge" wasted space,
given a certain usage pattern.  Consider a 4K 'count' with a, say, 3K 'wasted'.
I think that will go the "shared" path and waste 3K. I know people who would
say 'Ouch' (in French or Anglo-Saxon) to that.

As mentioned, I suspect, but have no data, that a simple `if (wasted > MAGIC_NUMBER)` would work here, over a wide range of guessed "MAGIC_NUMBER`s, each suitably explained.

It is a minor issue, but I do not see how  INITIAL_CAPACITY comes into play?
It is less of a 'magic number' both in mystery and magnitude than 256 but I miss
the logical connection, if any.

As David & I have discussed in a couple of other places, Discourse & Issues here,
the whole 'wasted' discussion is intimately wrapped up in the "grow-the-buffer" algorithm,
especially the small size of the first few (5? or so) allocations. That discussion
should not hold up progress here, but needs to be kept in the background.
I think the "Historical 1.5 grow-the-buffer" algorithm allocates too little at the
beginning and too much as the buffer grows larger.  Wild guesses.

Reply:
s/ each suitably explained./the finally chosen one suitably explained/

Reply:
Thank you @LeeTibbert for your review.
I'll try to address the different comments/questions one by one.

>The 256 'magic number' in the original code was never explained.

I'm not a big fan of it neither, as it makes the code a bit cryptic.
However, after a deeper analysis, we may find a good reason for keeping or modifying this value.

>I suspect that this line is problematic.
>By my, possibly flawed reading, this line of code will allow "huge" wasted space,
>given a certain usage pattern. Consider a 4K 'count' with a, say, 3K 'wasted'.
>I think that will go the "shared" path and waste 3K. 

I may be wrong but I don't think it will lead to a wasted backing Array ("shared" path) with the data provided in your example.
I summarized your data in a spreadsheet, where I computed additional values as follows:

| Variable | Value |
| ------------- | ------------- |
| count  | 4000  |
| wasted  | 3000  |
| length | 7000 |
| wasted_limit	| 2002 |
| wasted > wasted_limit ("sharing" disabled)	| true |

If I got your example, and since `val wasted = value.length - count`, the backing Array length should 7K, but this is not needed for further calculations.
And the resulting wasted_limit would be estimated as `(count / 2) + 2` which equals here 2002.
Thus the sharing test will fail, and the used chars will be copied in the created String.

>It is a minor issue, but I do not see how INITIAL_CAPACITY comes into play?
It is less of a 'magic number' both in mystery and magnitude than 256 but I miss
the logical connection, if any.

The idea is that if the waste is very small (lower than 16), we thus always enable sharing to avoid additional allocations.
But you are right, it's still a magic value and it could be questioned.
The goal behind this PR, is to set the number of `new Array()/new String()` to the bare minimum, without affecting _too much_ the wasted memory in the underlying backing Array. The rationale, is that less allocations means less garbage and thus less work for the GC.
The **best tradeoff** between less allocations and less waste in the backing Array might be tricky to find. The current PR proposes one possible model. Also important to note: what makes me a bit confident about this model, is that it has been adopted in Harmony and Android reference implementations.

>the whole 'wasted' discussion is intimately wrapped up in the "grow-the-buffer" algorithm,
especially the small size of the first few (5? or so) allocations. That discussion
should not hold up progress here, but needs to be kept in the background.
I think the "Historical 1.5 grow-the-buffer" algorithm allocates too little at the
beginning and too much as the buffer grows larger. Wild guesses.

Totally right, there might be room for improvement regarding the "grow-the-buffer" algorithm.
But this can be treated independently IMO.

Reply:
> But this can be treated independently IMO
Agreed

Thank you for the explanations.   

Discussion, not to block this PR or burden you.
If it does not give away proprietary details, may I ask:

* It sounds like you have tried this modification in your work.
   What did you see?

* Do you reuse a StringBuilder or is it single use?

* Do you add a lot of small strings to that SB or a few large (>= 1K) strings.

* Any idea of the expected lifetime of the child Strings?  Minutes, hours, days?

If I were to take a run at this, which I can not do in the near future,
I think I would first instrument a base load, and then run one of the
the Scala benchmarks against it.  I do not know if there is a String
intensive one.

I think I would try to set the lower acceptable bound higher (say 32)
to try to minimize allocations, and then the higher bound much lower,
maybe (>>2 + mumble).  The hope would be that, amortized, the
memory saved with "moderate" and "large" Strings would pay back
the memory spent on small strings.

Problem is that all of this almost certainly depends on the workload and
there does not appear to me to be an economic way of automatically 
adjusting the workload ( recent past behavior is single best predictor of future behavior.
Perhaps maintaining a running "max" and extending by some factor of that).
As you say, a problem for another issue. 

* Seems like being willing to "waste" 2001 bytes is a gift to the people
   who manufacture RAM.